File: python/src/wgrok/sender.py
"""WgrokSender - wraps payload in echo protocol and sends via configured platform."""

from __future__ import annotations

import aiohttp

from . import codec
from .config import SenderConfig
from .logging import get_logger
from .platform import platform_send_card, platform_send_message
from .protocol import ECHO_PREFIX, PAUSE_CMD, RESUME_CMD, format_echo, format_flags

PLATFORM_LIMITS = {
    "webex": 7439,
    "slack": 4000,
    "discord": 2000,
    "irc": 400,
}


class WgrokSender:
    def __init__(self, config: SenderConfig) -> None:
        self._config = config
        self._session: aiohttp.ClientSession | None = None
        self._logger = get_logger(config.debug, "wgrok.sender")
        self._paused = False
        self._buffer: list[tuple] = []

    async def _ensure_session(self) -> aiohttp.ClientSession:
        if self._session is None:
            self._session = aiohttp.ClientSession()
        return self._session

    async def send(
        self,
        payload: str,
        card: dict | None = None,
        compress: bool = False,
        from_slug: str | None = None,
    ) -> dict | list[dict]:
        """Format payload as echo message and send to the configured target.

        Args:
            payload: Text payload to send.
            card: Optional adaptive card JSON to attach.
            compress: If True, gzip+base64 encode the payload.
            from_slug: Sender identifier (defaults to config slug).
        """
        if self._paused:
            self._buffer.append((payload, card, compress, from_slug))
            self._logger.info("Buffered message (sender paused)")
            return {"buffered": True}

        session = await self._ensure_session()
        from_slug = from_slug or self._config.slug
        encrypted = self._config.encrypt_key is not None

        if compress:
            payload = codec.compress(payload)

        if encrypted:
            payload = codec.encrypt(payload, self._config.encrypt_key)

        flags = format_flags(compress, encrypted)
        text = format_echo(self._config.slug, from_slug, flags, payload)
        platform = self._config.platform
        token = self._config.webex_token
        target = self._config.target
        limit = PLATFORM_LIMITS.get(platform, 7439)

        if len(text.encode("utf-8")) > limit and card is None:
            # Auto-chunk: calculate overhead and split payload
            overhead = (
                len(ECHO_PREFIX.encode("utf-8"))
                + len(self._config.slug.encode("utf-8"))
                + len(from_slug.encode("utf-8"))
                + 3  # +3 for three colons in v2 format
            )
            max_payload = limit - overhead
            chunks = codec.chunk(payload, max_payload)
            self._logger.info(
                f"Payload exceeds {limit}B limit, sending {len(chunks)} chunks to {target} via {platform}"
            )
            results = []
            for i, ch in enumerate(chunks):
                chunk_flags = format_flags(compress, encrypted, i + 1, len(chunks))
                chunk_text = format_echo(self._config.slug, from_slug, chunk_flags, ch)
                results.append(await platform_send_message(platform, token, target, chunk_text, session))
            return results

        self._logger.info(f"Sending to {target} via {platform}: {text}")
        if card is not None:
            self._logger.info("Including card/rich content attachment")
            return await platform_send_card(platform, token, target, text, card, session)
        return await platform_send_message(platform, token, target, text, session)

    async def pause(self, notify: bool = True) -> None:
        """Pause sending — buffer all subsequent send() calls locally.

        Args:
            notify: If True (default), send ./pause to the router first.
                    Set to False when pause is router-initiated.
        """
        if notify:
            session = await self._ensure_session()
            await platform_send_message(
                self._config.platform, self._config.webex_token,
                self._config.target, PAUSE_CMD, session,
            )
        self._paused = True
        self._logger.info("Sender paused, buffering messages")

    async def resume(self, notify: bool = True) -> None:
        """Resume sending — flush buffered messages.

        Args:
            notify: If True (default), send ./resume to the router first.
                    Set to False when resume is router-initiated.
        """
        self._paused = False
        if notify:
            session = await self._ensure_session()
            await platform_send_message(
                self._config.platform, self._config.webex_token,
                self._config.target, RESUME_CMD, session,
            )
        buffered = self._buffer[:]
        self._buffer.clear()
        for p, c, comp, fs in buffered:
            await self.send(p, c, comp, fs)
        self._logger.info(f"Sender resumed, flushed {len(buffered)} message(s)")

    async def close(self) -> None:
        """Clean up the HTTP session."""
        if self._session:
            await self._session.close()
            self._session = None

File: go/sender.go
package wgrok

import (
	"fmt"
	"net/http"
	"sync"

	wmh "github.com/3rg0n/webex-message-handler/go"
)

// PlatformLimits defines message size limits per platform.
var PlatformLimits = map[string]int{
	"webex":   7439,
	"slack":   4000,
	"discord": 2000,
	"irc":     400,
}

// bufferedSend represents a message buffered during pause.
type bufferedSend struct {
	payload  string
	card     interface{}
	compress bool
}

// WgrokSender wraps payloads in the echo protocol and sends via Webex.
type WgrokSender struct {
	config  *SenderConfig
	client  *http.Client
	logger  wmh.Logger
	paused  bool
	buffer  []bufferedSend
	pauseMu sync.Mutex
}

// NewSender creates a new WgrokSender.
func NewSender(config *SenderConfig) *WgrokSender {
	return &WgrokSender{
		config: config,
		client: &http.Client{},
		logger: GetLogger(config.Debug, "wgrok.sender"),
	}
}

// Send formats payload as an echo message and sends to the configured target.
// If card is non-nil, it is attached as an adaptive card.
// If compress is true, the payload is gzip+base64 encoded before sending.
func (s *WgrokSender) Send(payload string, card interface{}) (map[string]interface{}, error) {
	return s.SendWithOptions(payload, card, false)
}

// SendWithOptions is like Send but with an explicit compress flag.
func (s *WgrokSender) SendWithOptions(payload string, card interface{}, compress bool) (map[string]interface{}, error) {
	s.pauseMu.Lock()
	if s.paused {
		if len(s.buffer) >= 1000 {
			s.logger.Warn("Pause buffer full (1000), dropping oldest message")
			s.buffer = s.buffer[1:]
		}
		s.buffer = append(s.buffer, bufferedSend{
			payload:  payload,
			card:     card,
			compress: compress,
		})
		s.pauseMu.Unlock()
		s.logger.Info("Sender is paused, buffering message")
		return map[string]interface{}{"buffered": true}, nil
	}
	s.pauseMu.Unlock()

	encrypted := s.config.EncryptKey != nil

	if compress {
		encoded, err := Compress(payload)
		if err != nil {
			return nil, fmt.Errorf("compress payload: %w", err)
		}
		payload = encoded
	}

	if encrypted {
		encoded, err := Encrypt(payload, s.config.EncryptKey)
		if err != nil {
			return nil, fmt.Errorf("encrypt payload: %w", err)
		}
		payload = encoded
	}

	// Use slug as both to and from (for sender context)
	to := s.config.Slug
	from := s.config.Slug
	flags := FormatFlags(compress, encrypted, 0, 0)

	text := FormatEcho(to, from, flags, payload)
	limit, ok := PlatformLimits[s.config.Platform]
	if !ok {
		limit = 7439
	}
	if len([]byte(text)) > limit && card == nil {
		// Estimate overhead for chunked format
		// Worst case for flags in chunking: "ze999/999" = 9 chars, we'll use 11 for safety
		flagOverhead := 11
		overhead := len([]byte(EchoPrefix)) + len([]byte(to)) + 1 + len([]byte(from)) + 1 + flagOverhead + 1
		maxPayload := limit - overhead
		chunks, err := Chunk(payload, maxPayload)
		if err != nil {
			return nil, fmt.Errorf("chunk payload: %w", err)
		}
		s.logger.Info(fmt.Sprintf("Payload exceeds %dB limit, sending %d chunks to %s", limit, len(chunks), s.config.Target))
		var lastResult map[string]interface{}
		for i, ch := range chunks {
			chunkFlags := FormatFlags(compress, encrypted, i+1, len(chunks))
			chunkText := FormatEcho(to, from, chunkFlags, ch)
			result, err := PlatformSendMessage(s.config.Platform, s.config.WebexToken, s.config.Target, chunkText, s.client)
			if err != nil {
				return nil, err
			}
			lastResult = result
		}
		return lastResult, nil
	}
	s.logger.Info(fmt.Sprintf("Sending to %s: %s", s.config.Target, text))
	if card != nil {
		s.logger.Info("Including adaptive card attachment")
		return PlatformSendCard(s.config.Platform, s.config.WebexToken, s.config.Target, text, card, s.client)
	}
	return PlatformSendMessage(s.config.Platform, s.config.WebexToken, s.config.Target, text, s.client)
}

// Pause pauses message delivery. If notify is true, sends a pause control message to the target.
func (s *WgrokSender) Pause(notify bool) error {
	s.pauseMu.Lock()
	s.paused = true
	s.pauseMu.Unlock()

	if notify {
		s.logger.Info(fmt.Sprintf("Sending pause notification to %s", s.config.Target))
		_, err := PlatformSendMessage(s.config.Platform, s.config.WebexToken, s.config.Target, PauseCmd, s.client)
		if err != nil {
			return fmt.Errorf("send pause message: %w", err)
		}
	}
	return nil
}

// Resume resumes message delivery. If notify is true, sends a resume control message to the target.
// Then flushes any buffered messages.
func (s *WgrokSender) Resume(notify bool) error {
	s.pauseMu.Lock()
	s.paused = false
	buffer := s.buffer
	s.buffer = []bufferedSend{}
	s.pauseMu.Unlock()

	if notify {
		s.logger.Info(fmt.Sprintf("Sending resume notification to %s", s.config.Target))
		_, err := PlatformSendMessage(s.config.Platform, s.config.WebexToken, s.config.Target, ResumeCmd, s.client)
		if err != nil {
			return fmt.Errorf("send resume message: %w", err)
		}
	}

	s.logger.Info(fmt.Sprintf("Resumed, flushing %d buffered message(s)", len(buffer)))
	for _, buffSend := range buffer {
		_, err := s.SendWithOptions(buffSend.payload, buffSend.card, buffSend.compress)
		if err != nil {
			s.logger.Error(fmt.Sprintf("Failed to send buffered message: %v", err))
		}
	}
	return nil
}
